### ISSUE TITLE
kube-scheduler gets stuck if there's a pod in a namespace that doesn't exist assigned to a node that doesn't exist

### ISSUE THREAD

**Issue by mikebryant** [kubernetes/kubernetes#43806]
Originally discussed here: https://kubernetes.slackarchive.io/kubernetes-users/page-100/ts-1490792092215676

**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): FailedScheduling "not found"

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): BUG REPORT

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"6+", GitVersion:"v1.6.0-beta.3", GitCommit:"0cd5ed469508e6dfc807ee6681561c845828917e", GitTreeState:"clean", BuildDate:"2017-03-11T04:20:22Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.2+coreos.1", GitCommit:"3ed7d0f453a5517245d32a9c57c39b946e578821", GitTreeState:"clean", BuildDate:"2017-01-18T01:43:45Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**: CoreOS
- **Cloud provider or hardware configuration**: Openstack
- **OS** (e.g. from /etc/os-release):  `Container Linux by CoreOS 1235.9.0 (Ladybug)`
- **Kernel** (e.g. `uname -a`): `Linux kubernetes-cit-kubernetes-cr2-2-1490788943 4.7.3-coreos-r2 #1 SMP Thu Feb 2 02:26:10 UTC 2017 x86_64 Intel Xeon E312xx (Sandy Bridge) GenuineIntel GNU/Linux`
- **Install tools**: custom/self-hosted
- **Others**:

**What happened**:
After a rolling upgrade in openstack (i.e. a complete replacement, all of the old nodes were deleted), the scheduler stopped being able to schedule *any* pods:
```
E0329 12:50:13.640534       1 factory.go:583] Error scheduling kube-system kube-scheduler-1480708505-8jrbk: node 'kubernetes-cit-kubernetes-cr1-0-1490206280' not found; retrying
I0329 12:50:13.640610       1 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"kube-scheduler-1480708505-8jrbk", UID:"fec110b8-1477-11e7-ab4c-fa163e1913d3", APIVersion:"v1", ResourceVersion:"59476545", FieldPath:""}): type: 'Warning' reason: 'FailedScheduling' node 'kubernetes-cit-kubernetes-cr1-0-1490206280' not found
E0329 12:50:13.742129       1 factory.go:583] Error scheduling log-aggregation kafka-server-0: node 'kubernetes-cit-kubernetes-cr1-0-1490206280' not found; retrying
I0329 12:50:13.742224       1 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"log-aggregation", Name:"kafka-server-0", UID:"4264d9c5-1474-11e7-9695-fa163e75820f", APIVersion:"v1", ResourceVersion:"59469412", FieldPath:""}): type: 'Warning' reason: 'FailedScheduling' node 'kubernetes-cit-kubernetes-cr1-0-1490206280' not found
E0329 12:50:13.746350       1 factory.go:583] Error scheduling kube-system kube-scheduler-1480708505-bx4lq: node 'kubernetes-cit-kubernetes-cr1-0-1490206280' not found; retrying
I0329 12:50:13.746411       1 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"kube-scheduler-1480708505-bx4lq", UID:"62c3344f-1476-11e7-ab4c-fa163e1913d3", APIVersion:"v1", ResourceVersion:"59466732", FieldPath:""}): type: 'Warning' reason: 'FailedScheduling' node 'kubernetes-cit-kubernetes-cr1-0-1490206280' not found
E0329 12:50:13.748522       1 factory.go:583] Error scheduling kube-system kube-controller-manager-2028335071-n43xs: node 'kubernetes-cit-kubernetes-cr1-0-1490206280' not found; retrying
I0329 12:50:13.748616       1 event.go:217] Event(api.ObjectReference{Kind:"Pod", Namespace:"kube-system", Name:"kube-controller-manager-2028335071-n43xs", UID:"5db1d4f0-1476-11e7-ab4c-fa163e1913d3", APIVersion:"v1", ResourceVersion:"59466119", FieldPath:""}): type: 'Warning' reason: 'FailedScheduling' node 'kubernetes-cit-kubernetes-cr1-0-1490206280' not found
```

**What you expected to happen**:
This to work

**How to reproduce it** (as minimally and precisely as possible):
Guess (not tried):
- Create a namespace, create a pod in that namespace, have it scheduled to a node.
- Delete the namespace without the pod being deleted (User report: was `kubectl delete namespace log-management`) - not sure why this didn't delete everything
- Now delete that node

**Anything else we need to know**:
Tried this: `etcdctl ls /registry --recursive | grep -v events | xargs -n1 -i etcdctl get {} | grep kubernetes-cit-kubernetes-cr1-0-1490206280` - to hopefully find any references to that node anywhere, and the only thing that came back was:
```
{"kind":"Pod","apiVersion":"v1","metadata":{"name":"cerebro-2756361696-4nlhx","generateName":"cerebro-2756361696-","namespace":"log-management","selfLink":"/api/v1/namespaces/log-management/pods/cerebro-2756361696-4nlhx","uid":"e178ad25-10bc-11e7-96a2-fa163e75820f","creationTimestamp":"2017-03-24T18:08:28Z","labels":{"appId":"logmanagement","component":"cerebro","pod-template-hash":"2756361696"},"annotations":{"kubernetes.io/created-by":"{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicaSet\",\"namespace\":\"log-management\",\"name\":\"cerebro-2756361696\",\"uid\":\"9327d1c0-098f-11e7-ae04-fa163ed5ffed\",\"apiVersion\":\"extensions\",\"resourceVersion\":\"57124035\"}}\n"},"ownerReferences":[{"apiVersion":"extensions/v1beta1","kind":"ReplicaSet","name":"cerebro-2756361696","uid":"9327d1c0-098f-11e7-ae04-fa163ed5ffed","controller":true}]},"spec":{"volumes":[{"name":"default-token-pmt34","secret":{"secretName":"default-token-pmt34","defaultMode":420}}],"containers":[{"name":"cerebro","image":"mirror-internal.docker.tech.lastmile.com/log-aggregation/cerebro-docker:b2f1b3871bde45c2f97072aa7ec73296bbde03e5","args":["-Dconfig.file=/etc/cerebro/cerebro.conf","-v"],"ports":[{"containerPort":9000,"protocol":"TCP"}],"env":[{"name":"CEREBRO_SECRET","valueFrom":{"secretKeyRef":{"name":"cerebro-secret","key":"play.crypto.secret"}}}],"resources":{},"volumeMounts":[{"name":"default-token-pmt34","readOnly":true,"mountPath":"/var/run/secrets/kubernetes.io/serviceaccount"}],"terminationMessagePath":"/dev/termination-log","imagePullPolicy":"Always"}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","serviceAccountName":"default","serviceAccount":"default","nodeName":"kubernetes-cit-kubernetes-cr1-0-1490206280","securityContext":{},"imagePullSecrets":[{"name":"registry.docker.tech.lastmile.com"}]},"status":{"phase":"Running","conditions":[{"type":"Initialized","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-03-24T18:08:28Z"},{"type":"Ready","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-03-24T18:08:41Z"},{"type":"PodScheduled","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-03-24T18:08:28Z"}],"hostIP":"10.33.234.162","podIP":"172.31.84.15","startTime":"2017-03-24T18:08:28Z","containerStatuses":[{"name":"cerebro","state":{"running":{"startedAt":"2017-03-24T18:08:41Z"}},"lastState":{},"ready":true,"restartCount":0,"image":"mirror-internal.docker.tech.lastmile.com/log-aggregation/cerebro-docker:b2f1b3871bde45c2f97072aa7ec73296bbde03e5","imageID":"docker-pullable://mirror-internal.docker.tech.lastmile.com/log-aggregation/cerebro-docker@sha256:2adc458073b54df6541a034c81f0e8574766f320274b441d8a58c0e347e75a88","containerID":"docker://236251993dbe2612add624605189ae74dadfe83706bf8d70327441176bee6067"}]}}
```
Trying to delete that pod: `Error from server (NotFound): pods "cerebro-2756361696-4nlhx" not found`
This situation was resolved by recreating the namespace the pod used to be in - then the pod was immediately deleted (either by the node controller or the replicaset controller, unsure), and the scheduler started working again

**Comment by jayunit100** [comment]
cc @kubernetes/sig-scheduling-bugs

**Comment by k82cn** [comment]
Any logs from controller-manager? I'd like to know what's happened when namespace deleted.

**Comment by mikebryant** [comment]
I'm afraid not - we have very limited log storage, and the namespace deletion happened something like a week prior.

**Comment by timothysc** [comment]
> Delete the namespace without the pod being deleted (User report: was kubectl delete namespace log-management) - not sure why this didn't delete everything 

This should cleanup all the pods... So now I have more questions

- Is this a timing artifact?  How long did you wait before you deleted the node? 
- Is the pod stuck in "Terminating"? 
- Did you try re-starting the scheduler?

**Comment by mikebryant** [comment]
The namespace was deleted around a week before the node was. (I became aware of this only during debugging why everything was broken, I didn't delete the namespace myself)


The pod wasn't visible via kubectl at all - indeed, trying to delete it via kubectl claimed that it didn't exist.
When I created the namespace again (in an attempt to allow deletion via kubectl) by the time I checked something had deleted it, I didn't think to figure out exactly which component then deleted it. And I didn't see what the api thought of it before that happened.
However, in the json blob above I pulled out of etcd, it looks like `status.phase == Running`


Yup, tried restarting the scheduler, and all of the apiservers, and then the scheduler again

**Comment by davidopp** [comment]
@bsalamat

**Comment by k82cn** [comment]
>Delete the namespace without the pod being deleted (User report: was kubectl delete namespace log-management) - not sure why this didn't delete everything
>This should cleanup all the pods... So now I have more questions

Did not test; but I'm thinking what will happen if controller-manager abort/crash when before deleting all pods; will it get namespace deleted notification when start-up next time?

**Comment by timothysc** [comment]
>  but I'm thinking what will happen if controller-manager abort/crash before deleting all pods

The delete is direct in the api-server and the only thing that lingers that requires rectification are pods stuck in "Terminating"

**Comment by k82cn** [comment]
> The delete is direct in the api-server and the only thing that lingers that requires rectification are pods stuck in "Terminating"

@timothysc , thanks very much; just try it in local and check the code: the ns will be in Terminating state until all related resources are deleted; so even the controller manager abort/crash, it'll continue to delete Pods when it's re-started.

**Comment by timothysc** [comment]
This points back to our legacy namespace delete issue, which we can/should fix in 1.7 

/cc @wojtek-t @gmarek

**Comment by wojtek-t** [comment]
> This points back to our legacy namespace delete issue, which we can/should fix in 1.7

@timothysc - I'm not sure I know what you are referring to. BTW - do we understand what has happened here?

**Comment by mikebryant** [comment]
/remove-lifecycle rotten

@timothysc was this ever fixed?

**Comment by bsalamat** [comment]
@mikebryant I am not so sure if the deletion of the namespace had a role in this bug. It looks to me that this bug is a dup of  #56057 which led us to #56261. Then, #56622 was sent to address the issue. So, I believe this bug is fixed in 1.9.

**Comment by bsalamat** [comment]
The fix is back ported to 1.8 as well: #58038.
